x/vesting/continuous_vesting_account.go

```diff
@@ -0,0 +1,190 @@
+	*account_abstractionv1.MsgExecuteResponse, error,
+) {
+	return cva.BaseVesting.ExecuteMessages(ctx, msg, cva.GetVestingCoins)
+}
+
+// --------------- Query -----------------
+
+// GetVestedCoins returns the total number of vested coins. If no coins are vested,
+// nil is returned.
+func (cva ContinuousVestingAccount) GetVestedCoins(ctx context.Context, blockTime time.Time) (sdk.Coins, error) {
+	vestedCoins := sdk.Coins{}
+
+	// We must handle the case where the start time for a vesting account has
+	// been set into the future or when the start of the chain is not exactly
+	// known.
+	startTime, err := cva.StartTime.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	endTime, err := cva.EndTime.Get(ctx)
+	if err != nil {
+		return nil, err
+	}
+	var originalVesting sdk.Coins
+	cva.IterateCoinEntries(ctx, cva.OriginalVesting, func(key string, value math.Int) (stop bool) {
+		originalVesting = append(originalVesting, sdk.NewCoin(key, value))
+		return false
+	})
+	if startTime.GTE(math.NewInt(blockTime.Unix())) {
+		return vestedCoins, nil
+	} else if endTime.LTE(math.NewInt(blockTime.Unix())) {
+		return originalVesting, nil
+	}
+
+	// calculate the vesting scalar
+	x := math.NewInt(blockTime.Unix()).Sub(startTime).Int64()
+	y := endTime.Sub(startTime).Int64()
+	s := math.LegacyNewDec(x).Quo(math.LegacyNewDec(y))
+
+	for _, ovc := range originalVesting {
+		vestedAmt := math.LegacyNewDecFromInt(ovc.Amount).Mul(s).RoundInt()
+		vestedCoins = append(vestedCoins, sdk.NewCoin(ovc.Denom, vestedAmt))
+	}
+
+	return vestedCoins, nil
+}
+
+// GetVestingCoins returns the total number of vesting coins. If no coins are
+// vesting, nil is returned.
+func (cva ContinuousVestingAccount) GetVestingCoins(ctx context.Context, blockTime time.Time) (sdk.Coins, error) {
+	originalVesting := sdk.Coins{}
+	cva.IterateCoinEntries(ctx, cva.OriginalVesting, func(key string, value math.Int) (stop bool) {
+		originalVesting = append(originalVesting, sdk.NewCoin(key, value))
+		return false
+	})
+	vestedCoins, err := cva.GetVestedCoins(ctx, blockTime)
+	if err != nil {
+		return nil, err
+	}
+	return originalVesting.Sub(vestedCoins...), nil
+}
+
+func (cva ContinuousVestingAccount) QueryVestedCoins(ctx context.Context, msg *vestingtypes.QueryVestedCoinsRequest) (
+	*vestingtypes.QueryVestedCoinsResponse, error,
+) {
+	hs := cva.headerService.GetHeaderInfo(ctx)
+	vestedCoins, err := cva.GetVestedCoins(ctx, hs.Time)
+	if err != nil {
+		return nil, err
+	}
+	return &vestingtypes.QueryVestedCoinsResponse{
+		VestedVesting: vestedCoins,
+	}, nil
+}
+
+func (cva ContinuousVestingAccount) QueryVestingCoins(ctx context.Context, msg *vestingtypes.QueryVestingCoinsRequest) (
+	*vestingtypes.QueryVestingCoinsResponse, error,
+) {
+	hs := cva.headerService.GetHeaderInfo(ctx)
+	vestingCoins, err := cva.GetVestingCoins(ctx, hs.Time)
+	if err != nil {
+		return nil, err
+	}
+	return &vestingtypes.QueryVestingCoinsResponse{
+		VestingCoins: vestingCoins,
+	}, nil
+}
```

Comment:
should we combine this into a single query. In a frontend they would most likely want to show both values, so separate queries means two queries

Reply:
yeah agreed since it kinda repeative